devices/src/virtio/block.rs

```diff
@@ -166,20 +166,46 @@ impl Request {
             return Err(Error::UnexpectedWriteOnlyDescriptor);
         }
 
-        let request_type = request_type(&mem, avail_desc.addr)?;
-        let sector = sector(&mem, avail_desc.addr)?;
+        let mut req = Request {
+            request_type: request_type(&mem, avail_desc.addr)?,
+            sector: sector(&mem, avail_desc.addr)?,
+            data_addr: GuestAddress(0),
+            data_len: 0,
+            status_addr: GuestAddress(0),
+        };
+
+        if req.request_type == RequestType::Flush {
```

Comment:
There should be a way to refactor this to not have to duplicate the code/logic. We can discuss suggestions offline.